llm-chain-local/src/executor.rs

```diff
@@ -0,0 +1,195 @@
+        let mut output = String::new();
+        session
+            .infer::<std::convert::Infallible>(
+                self.llm.as_ref(),
+                prompt.to_text().as_str(),
+                // EvaluateOutputRequest
+                &mut Default::default(),
+                &mut rand::thread_rng(),
+                |t| {
+                    output.push_str(t);
+
+                    Ok(())
+                },
+            )
+            .map_err(|e| Error::InnerError(Box::new(e)))?;
+
+        Ok(output.into())
+    }
+
+    fn tokens_used(
+        &self,
+        options: Option<&Self::PerInvocationOptions>,
+        prompt: &llm_chain::prompt::Prompt,
+    ) -> Result<llm_chain::tokens::TokenCount, llm_chain::tokens::PromptTokensError> {
+        let tokenizer = self.get_tokenizer(options)?;
+        let input = prompt.to_text();
+
+        let tokens_used = tokenizer
+            .tokenize_str(&input)
+            .map_err(|_e| PromptTokensError::UnableToCompute)?
+            .len() as i32;
+        let max_tokens = self.max_tokens_allowed(options);
+        Ok(llm_chain::tokens::TokenCount::new(max_tokens, tokens_used))
+    }
+
+    fn max_tokens_allowed(&self, _: Option<&Self::PerInvocationOptions>) -> i32 {
+        self.llm.n_context_tokens().try_into().unwrap_or(2048)
+    }
+
+    fn get_tokenizer(
+        &self,
+        _: Option<&Self::PerInvocationOptions>,
+    ) -> Result<Self::StepTokenizer<'_>, llm_chain::tokens::TokenizerError> {
+        Ok(LocalLlmTokenizer::new(self))
+    }
+
+    fn get_text_splitter(
+        &self,
+        _: Option<&Self::PerInvocationOptions>,
+    ) -> Result<Self::TextSplitter<'_>, Self::Error> {
+        Ok(LocalLlmTextSplitter::new(self))
+    }
+}
+
+pub struct LocalLlmTokenizer<'a> {
+    llm: &'a dyn llm::Model,
+}
+
+impl<'a> LocalLlmTokenizer<'a> {
+    pub fn new(executor: &'a Executor) -> Self {
+        LocalLlmTokenizer {
+            llm: executor.llm.as_ref(),
+        }
+    }
+}
+
+impl Tokenizer<llm::TokenId> for LocalLlmTokenizer<'_> {
+    fn tokenize_str(&self, doc: &str) -> Result<Vec<llm::TokenId>, TokenizerError> {
+        match &self.llm.vocabulary().tokenize(doc, false) {
+            Ok(tokens) => Ok(tokens.into_iter().map(|t| t.1).collect()),
+            Err(_) => Err(TokenizerError::TokenizationError),
+        }
+    }
+
+    fn to_string(&self, tokens: Vec<llm::TokenId>) -> Result<String, TokenizerError> {
+        let mut res = String::new();
+        let mut token_utf8_buf = llm::TokenUtf8Buffer::new();
+        for token_id in tokens {
+            // Buffer the token until it's valid UTF-8, then call the callback.
+            if let Some(tokens) =
+                token_utf8_buf.push(self.llm.vocabulary().token(token_id as usize))
+            {
+                res.push_str(&tokens)
+            }
+        }
+
+        Ok(res)
+    }
+}
+
+pub fn load<M: llm::KnownModel + 'static>(
```

Comment:
Is this one needed?

Reply:
Which one? The lifetime? If I remove the lifetime, it breaks and says the parameter M may not live long enough...

Reply:
Ah meant the whole function :)

Reply:
Ah, it doesn't need to be published, but it is currently being used.